heron/stmgr/src/cpp/manager/stateful-restorer.cpp

```diff
@@ -0,0 +1,260 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "manager/stateful-restorer.h"
+#include <functional>
+#include <iostream>
+#include <list>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+#include "manager/stmgr-server.h"
+#include "manager/ckptmgr-client.h"
+#include "manager/stmgr-clientmgr.h"
+#include "util/tuple-cache.h"
+#include "metrics/metrics.h"
+#include "proto/messages.h"
+#include "basics/basics.h"
+#include "errors/errors.h"
+#include "threads/threads.h"
+#include "network/network.h"
+
+namespace heron {
+namespace stmgr {
+
+// Stats for restore
+const sp_string METRIC_START_RESTORE = "__start_restore";
+const sp_string METRIC_START_RESTORE_IN_PROGRESS = "__start_restore_in_progress";
+const sp_string METRIC_START_RESTORE_FAILED = "__start_restore_failed";
+const sp_string METRIC_CKPT_REQUESTS = "__ckpt_requests";
+const sp_string METRIC_CKPT_RESPONSES = "__ckpt_responses";
+const sp_string METRIC_CKPT_RESPONSES_IGNORED = "__ckpt_responses_ignored";
+const sp_string METRIC_CKPT_RESPONSES_ERROR = "__ckpt_responses_error";
+const sp_string METRIC_INSTANCE_RESTORE_REQUESTS = "__instance_restore_requests";
+const sp_string METRIC_INSTANCE_RESTORE_RESPONSES = "__instance_restore_responses";
+const sp_string METRIC_INSTANCE_RESTORE_RESPONSES_IGNORED = "__instance_restore_response_ignored";
+
+StatefulRestorer::StatefulRestorer(CkptMgrClient* _ckptmgr,
+                             StMgrClientMgr* _clientmgr, TupleCache* _tuple_cache,
+                             StMgrServer* _server,
+                             common::MetricsMgrSt* _metrics_manager_client,
+                             std::function<void(proto::system::StatusCode,
+                                                std::string, sp_int64)> _restore_done_watcher) {
+  ckptmgr_ = _ckptmgr;
+  clientmgr_ = _clientmgr;
+  tuple_cache_ = _tuple_cache;
+  server_ = _server;
+
+  in_progress_ = false;
+  restore_done_watcher_ = _restore_done_watcher;
+  metrics_manager_client_ = _metrics_manager_client;
+  multi_count_metric_ = new common::MultiCountMetric();
+  time_spent_metric_ = new common::TimeSpentMetric();
+  metrics_manager_client_->register_metric("__stateful_restore_count", multi_count_metric_);
+  metrics_manager_client_->register_metric("__stateful_restore_time", time_spent_metric_);
+}
+
+StatefulRestorer::~StatefulRestorer() {
+  metrics_manager_client_->unregister_metric("__stateful_restore_count");
+  metrics_manager_client_->unregister_metric("__stateful_restore_time");
+  delete multi_count_metric_;
+  delete time_spent_metric_;
+}
+
+void StatefulRestorer::StartRestore(const std::string& _checkpoint_id, sp_int64 _restore_txid,
```

Comment:
Could you add comments to StartRestore, HandleCheckpointState and HandleInstanceRestoredState clarifying what the responsibilities of each is.